src/clj/fluree/db/vector/index_graph.cljc

```diff
@@ -0,0 +1,171 @@
+(def iri-property (str iri/f-idx-flatrank-ns "property"))
+(def iri-limit (str iri/f-idx-flatrank-ns "limit"))
+(def iri-id (str iri/f-idx-flatrank-ns "id"))
+(def iri-score (str iri/f-idx-flatrank-ns "score"))
+(def iri-vector (str iri/f-idx-flatrank-ns "vector"))
+(def iri-xsd-float "http://www.w3.org/2001/XMLSchema#float")
+
+(def metric-re (re-pattern (str iri/f-idx-flatrank-ns "(.*)")))
+
+(def result-sort (fn [a b] (compare (get a :score) (get b :score))))
+(def reverse-result-sort (fn [a b] (compare (get b :score) (get a :score))))
+
+(def metrics
+  {:dotproduct {::score-fn vector.score/dotproduct
+                ::sort-fn  reverse-result-sort}
+   :cosine     {::score-fn vector.score/cosine-similarity
+                ::sort-fn  reverse-result-sort}
+   :distance   {::score-fn vector.score/euclidian-distance
+                ::sort-fn  result-sort}})
+
+(defn- prop-iri
+  [triple]
+  (-> triple (nth 1) ::where/iri))
+
+(defn- obj-val
+  [triple solution]
+  (let [o (nth triple 2)]
+    (or (where/get-value o)
+        (->> (where/get-variable o)
+             (get solution)
+             (where/get-value)))))
+
+(defn- obj-var
+  [triple]
+  (-> triple (nth 2) ::where/var))
+
+(defn- obj-iri
+  [triple]
+  (-> triple (nth 2) ::where/iri))
+
+(defn extract-metric
+  "Takes the graph alias as a string and extracts the metric name from the
+  end of the IRI"
+  [graph-alias]
+  (some-> (re-find metric-re graph-alias)
+          second
+          str/lower-case
+          keyword))
+
+(defn parse-search-graph
+  [graph-alias solution graph-triples]
+  (try*
+    (let [metric (extract-metric graph-alias)]
+      (reduce
+       (fn [acc triple]
+         (cond
+           (= iri-compare (prop-iri triple))
+           (assoc acc ::compare (obj-val triple solution))
+
+           (= iri-property (prop-iri triple))
+           (assoc acc ::property (obj-iri triple))
+
+           (= iri-limit (prop-iri triple))
+           (assoc acc ::limit (obj-val triple solution))
+
+           (= iri-id (prop-iri triple))
+           (assoc-in acc [::result ::id] (obj-var triple))
+
+           (= iri-score (prop-iri triple))
+           (assoc-in acc [::result ::score] (obj-var triple))
+
+           (= iri-vector (prop-iri triple))
+           (assoc-in acc [::result ::vector] (obj-var triple))
+
+           :else
+           acc))
+       {::metric metric}
+       graph-triples))
+    (catch* e
+            (throw (ex-info (str "Unable to parse search graph provided for index: " graph-alias)
+                            {:status 400
+                             :error  :db/invalid-query} e)))))
+
+(defn extract-vectors
+  [{::keys [property]} db]
+  (let [pid (iri/encode-iri db property)]
+    ;; For now, pulling all matching values from full index
+    ;; once hitting the actual vector index, we'll only need
+    ;; to pull matches out of novelty (if that)
+    (query-range/index-range db :post = [pid])))
```

Comment:
If this is too large to fit into memory, what will happen?

Reply:
OOM Exception - which is the same thing that any query whose results won't fit into memory today will have happen, even a :limit 10 but with an order-by has the same risk, or aggregates over large results. So it isn't even large results that have this risk.

We've talk about disk spill-overs but right now that isn't something we have.